File: javascript/packages/fury/lib/writer/number.ts
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

const float32View = new Float32Array(1);
const int32View = new Int32Array(float32View.buffer);

export function toFloat16(value: number) {
  float32View[0] = value;
  const floatValue = int32View[0];
  const sign = (floatValue >>> 16) & 0x8000; // sign only
  const exponent = ((floatValue >>> 23) & 0xff) - 127; // extract exponent from floatValue
  const significand = floatValue & 0x7fffff; // extract significand from floatValue

  if (exponent === 128) { // floatValue is NaN or Infinity
    return sign | ((exponent === 128) ? 0x7c00 : 0x7fff);
  }

  if (exponent > 15) {
    return sign | 0x7c00; // return Infinity
  }

  if (exponent < -14) {
    return sign | 0x3ff; // returns ±max subnormal
  }

  if (exponent <= 0) {
    return sign | ((significand | 0x800000) >> (1 - exponent + 10));
  }

  return sign | ((exponent + 15) << 10) | (significand >> 13);
}

File: rust/fory-core/src/serializer/number.rs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

use crate::error::Error;
use crate::fory::Fory;
use crate::resolver::context::ReadContext;
use crate::resolver::context::WriteContext;
use crate::serializer::Serializer;
use crate::types::{FieldType, ForyGeneralList};

macro_rules! impl_num_serializer {
    ($name: ident, $ty:tt, $field_type: expr) => {
        impl Serializer for $ty {
            fn write(&self, context: &mut WriteContext) {
                context.writer.$name(*self);
            }

            fn read(context: &mut ReadContext) -> Result<Self, Error> {
                Ok(context.reader.$name())
            }

            fn reserved_space() -> usize {
                std::mem::size_of::<$ty>()
            }

            fn get_type_id(_fory: &Fory) -> i16 {
                ($field_type).into()
            }
        }
    };
}
impl ForyGeneralList for i8 {}
impl ForyGeneralList for u16 {}
impl ForyGeneralList for u32 {}
impl ForyGeneralList for u64 {}

impl_num_serializer!(i8, i8, FieldType::INT8);
impl_num_serializer!(u8, u8, FieldType::UINT8);
impl_num_serializer!(i16, i16, FieldType::INT16);
impl_num_serializer!(u16, u16, FieldType::UINT16);
impl_num_serializer!(i32, i32, FieldType::INT32);
impl_num_serializer!(u32, u32, FieldType::UINT32);
impl_num_serializer!(u64, u64, FieldType::UINT64);
impl_num_serializer!(i64, i64, FieldType::INT64);
impl_num_serializer!(f32, f32, FieldType::FLOAT);
impl_num_serializer!(f64, f64, FieldType::DOUBLE);
